x-pack/plugins/reporting/server/export_types/csv_searchsource/generate_csv/generate_csv.ts

```diff
@@ -0,0 +1,327 @@
+          params: {
+            ...index.fieldFormatMap[name]?.params,
+            timezone: settings.timezone,
+          },
+        };
+        index.setFieldFormat(name, format);
+      });
+    }
+
+    if (pitId) {
+      searchSource.setField('pit', { id: pitId, keep_alive: '2m' });
+    } else {
+      this.logger.error(`Unable to open a point in time with ${searchSource.getId()}!`);
+    }
+
+    do {
+      if (lastSortId) {
+        searchSource.setField('searchAfter', lastSortId);
+      }
+      const results = await searchSource.fetch();
+
+      if (results?.hits?.total == null) {
+        this.logger.error('Expected total number of records in the search response');
+        totalRecords = -1;
+      } else {
+        totalRecords = results.hits.total;
+      }
+
+      const hitsLength = results?.hits?.total || 0;
+      this.logger.debug(`Search results: ${hitsLength}`);
+
+      const table = tabifyDocs(results, index, { shallow: true, meta: true });
+      if (table.columns.length < 1) {
+        break; // empty report
+      }
+
+      // write the header and initialize formatters / column orderings
+      // depends on the table to know what order to place the columns
+      if (first) {
+        first = false;
+        this.generateHeader(fields, table, builder, settings);
+      }
+
+      if (table.rows.length < 1) {
+        break; // empty report with just the header
+      }
+
+      const formatters = this.getFormatters(table);
+      this.generateRows(fields, table, builder, formatters, settings);
+
+      // update last sort ID
+      lastSortId = results.hits.hits[results.hits.hits.length - 1]?.sort as EsQuerySearchAfter;
+      this.logger.debug(`last sort id: ${lastSortId}`);
+
+      // A query using a non-timebased index pattern needs the user to define
+      // sort field(s) in order to page the data using search_after and PIT
+      if (hitsLength && !lastSortId) {
+        throw new Error(
+          `No sort option used in the query: CSV export can not get all the data without a unique sort ID for every search result! Please choose one or more columns to sort by, and try again.`
+        );
+      }
+
+      // update iterator
+      currentRecord += table.rows.length;
+    } while (currentRecord < totalRecords - 1);
+
+    const size = builder.getSizeInBytes();
+    this.logger.debug(
+      `Finished generating. Total size in bytes: ${size}. Row count: ${this.csvRowCount}.`
+    );
+
+    if (this.csvContainsFormulas && escapeFormulaValues) {
+      warnings.push(
+        i18n.translate('xpack.reporting.exportTypes.csv.generateCsv.escapedFormulaValues', {
+          defaultMessage: 'CSV may contain formulas whose values have been escaped',
+        })
+      );
+    }
+
+    // clean PIT
+    if (pitId) {
+      this.logger.debug(`Closing PIT`);
+      await this.esClient.asCurrentUser.closePointInTime({ body: { id: pitId } });
+    }
```

Comment:
I'd recommend wrapping anything that can throw earlier in this function in a `try/catch`, so that you can still close the PIT if there's a failure. Otherwise you could have a PIT hanging around in your cluster unnecessarily until it hits the `keep_alive`. (For example line 290, or the call to `searchSource.fetch`)

Reply:
Thanks! I added 2adb303 for this